proposals/22-04-localize-command.md

```diff
@@ -0,0 +1,335 @@
+    │               └── kustomize
+    │                   └── v1.0.6
+    │                       └── examples
+    │                           └── helloWorld
+    │                               └── deployment.yaml
+    └── base
+        └── kustomization.yaml
+```
+
+After I run `kustomize localize example/overlay example new-space/localized-example`, I get the
+following:
+
+```
+├── example                         # old kustomization directory
+│   ├── overlay
+│   │   └── kustomization.yaml
+│   ├── localized-overlay
+│   │   ├── kustomization.yaml
+│   │   └── localized-files
+│   │       └── github.com
+│   │           └── kubernetes-sigs
+│   │               └── kustomize
+│   │                   └── v1.0.6
+│   │                       └── examples
+│   │                           └── helloWorld
+│   │                               └── deployment.yaml
+│   └── base
+│       └── kustomization.yaml
+└── new-space
+    └── localized-example            # the new, localized kustomization directory
+        ├── kustomization.yaml
+        ├── base
+        │   ├── kustomization.yaml
+        │   └── localized-files
+        │       └── github.com
+        │           └── kubernetes-sigs
+        │               └── kustomize
+        │                   └── v1.0.6
+        │                       └── examples
+        │                           ├── helloWorld
+        │                           │   └── configMap.yaml
+        │                           └── multibases
+        │                               ├── base
+        │                               │   └── pod.yaml
+        │                               ├── dev
+        │                               ├── kustomization.yaml
+        │                               ├── production
+        │                               └── staging
+        └── overlay
+            ├── kustomization.yaml
+            └── localized-files
+                └── github.com
+                    └── kubernetes-sigs
+                        └── kustomize
+                            └── v1.0.6
+                                └── examples
+                                    └── helloWorld
+                                        └── deployment.yaml
+```
+```
+# new-space/localized-example/overlay/kustomization.yaml
+resources:
+  - ../base
+  - ./localized-files/github.com/kubernetes-sigs/kustomize/examples/helloWorld/deployment.yaml
+```
+```
+# new-space/localized-example/base/kustomization.yaml
+resources:
+  - ./localized-files/github.com/kubernetes-sigs/kustomize/examples/multibases
+  - ./localized-files/github.com/kubernetes-sigs/kustomize/examples/helloWorld/configMap.yaml
+```
+
+Now, I upload `new-space/localized-example` from my local setup to my company’s internal package management site. I
+change the commands in my CI/CD pipeline to pull `localized-example` before running `kustomize build localized-example`,
+and the command executes successfully!
+
+### Risks and Mitigations
+N/A
```

Comment:
There's _always_ a risk of some kind. 😄 
If someone clairvoyant were to tell you there IS a problem, what are your guesses as to what it's about?

Some factors to consider addressing:
- Some of the referenced files may contain secrets (resources, or in an env file for example)
- Files localized may include executables controlled by third parties.
- The Kustomization tree being localized may reference bases in locations on the user's system that the person running the command does not expect (related to our scope arg and exclusion of loadrestrictor modifications)

Reply:
Thanks for the ideas! I want to make sure I'm understanding you correctly:
1. Referenced files may contain secrets: This would be a problem if either `localize` copies the file to a location with different read permissions or the user tried to upload the copy to a repo? 
   I'm thinking the command could identify secrets using `kind: Secret` and either
   * print warnings
   * throw an error unless a flag is specified to print warnings
   What do you think?
2. 3rd party executables: Natasha suggested we change the permissions of the executables after downloading to "not executable"

Reply:
> 3rd party executables: Natasha suggested we change the permissions of the executables after downloading to "not executable"

Elaborating on this; the purpose is that after running `kustomize localize`, the user would have to explicitly change the file permission of any downloaded exec files in order for them to execute.

Reply:
To clarify, we don't necessarily need to do anything about those factors in the design. I was just trying to brainstorm areas where we might be able to think of an associated risk to call out in this section.

1. I haven't actually managed to think up a way it could cause a problem or lead to novel exposure of the data in question. The user necessarily has permission to access it, or else the localize command would fail. There is a risk they will do something silly like localize secrets from a private remote repo and push the results somewhere insecure (e.g. maybe the target dir is a shared drive), but I don't think it's reasonable to expect us to try to warn against that.

2. This is an interesting idea for a safeguard. Right now, would the unlocalized `kustomize build` simply execute the files in question? I think it would (as long as the appropriate alpha plugin flags are set). If we changed the permissions for `build` as well though, we'd effectively be disabling remote exec functions. Ultimately, I think Catalog could remove this problem, since build in either context would require the exec function to be specifically allowlisted. But the more I think about it, the more I like your idea in the meantime.